src/app/services/audio.service.ts

```diff
@@ -5,177 +5,281 @@ import * as moment from 'moment';
+        this.updateStateEvents(event);
+        observer.next(event);
+      };
 
-    play() {
-        this.audioObj.play();
-    }
-
-    pause() {
+      this.addEvents(this.audioObj, this.audioEvents, handler);
+      return () => {
+        // Stop Playing
         this.audioObj.pause();
-    }
+        this.audioObj.currentTime = 0;
+        // remove event listeners
+        this.removeEvents(this.audioObj, this.audioEvents, handler);
+        // reset state
+        // this.resetState();
+      };
+    });
+  }
 
-    stop() {
-        this.stop$.next();
+  playStream(
+    { preview_url, id }: ITrack,
+    trackList?: ITrack[],
+    albumId?: string
+  ) {
+    this.audioID = id;
+    this.audioIDChange.next(id);
+    if (albumId) this.albumId = albumId;
+    if (trackList) {
+      this.trackList = trackList;
+      this.trackListChange.next(trackList);
+      this.lastRandom = [];
     }
+    this.streamObservable(preview_url)
+      .pipe(takeUntil(this.stop$))
+      .subscribe((event: Event) => {
+        if (event.type === 'ended') {
+          if (this.isAutorenewed) {
+            this.audioObj.currentTime = 0;
+            return this.play();
+          }
+          if (this.isShuffled) {
+            return this.playStream(this.trackList[this.randomize()]);
+          }
+          this.playNextTrack();
+        }
+      });
+  }
 
-    rewindTo(seconds: number) {
-        this.audioObj.currentTime = seconds;
-    }
+  private addEvents(obj, events, handler) {
+    events.forEach((event) => {
+      obj.addEventListener(event, handler);
+    });
+  }
 
-    playNextTrack() {
-        const currentTrack = this.trackList.find(track => track.id === this.audioID);
-        const nextTrack = this.trackList.find(track => track.track_number - 1 === currentTrack.track_number);
-        const isTrackListEnd = this.trackList.length === currentTrack.track_number + 1;
-        this.stop();
-        if (isTrackListEnd) return this.audioIDChange.next(null);
-        this.playStream(nextTrack);
-    }
+  private removeEvents(obj, events, handler) {
+    events.forEach((event) => {
+      obj.removeEventListener(event, handler);
+    });
+  }
 
-    formatTime(time: number, format: string = 'mm:ss') {
-        const momentTime = time * 1000;
-        return moment.utc(momentTime).format(format);
-    }
+  play() {
+    this.audioObj.play();
+  }
 
-    getState(): Observable<StreamState> {
-        return this.stateChange.asObservable();
-    }
+  pause() {
+    this.audioObj.pause();
+  }
 
-    getAudioID() {
-        return this.audioID;
-    }
+  stop() {
+    const volume = this.state.volume;
+    this.stop$.next();
+    this.state.volume = volume;
+  }
 
-    getAudioIDChange() {
-        return this.audioIDChange.asObservable();
-    }
+  rewindTo(seconds: number) {
+    this.audioObj.currentTime = seconds;
+  }
+
+  setVolume(volume: number) {
+    this.audioObj.volume = volume;
+    this.state.volume = volume;
```

Comment:
you can add additional event type to the listener - `volumechange`, and reassign value there.
https://developer.mozilla.org/ru/docs/Web/Guide/Events/Media_events